### DISCUSSION TITLE
A little help please?

### DISCUSSION THREAD

**Discussion by davemackintosh** [pest-parser/pest#949] [General]
I've been using ANTLR for a long while but over the last week I've been learning how to use Pest and Rust.

I've really enjoyed using it but I'm having an issue which I'm not sure how to fix.

Given the following, extracted from my grammar [here](https://github.com/elp-lang/elpc/blob/main/src/elp.pest) running against the "[kitchen sink](https://github.com/elp-lang/elpc/blob/main/examples/kitchen-sink.velp)"

```
FunctionCall = { Ident ~ (CallArgDeclarations | Block)+ }
Function = { FN ~ Ident ~ ParenArgDeclarations? ~ FunctionInherits? ~ ReturnType? ~ Block }

```

it doesn't appear to me that pest is able to distinguish between these two rules as I can't seem to get past 

```
Every 2.0s: echo ^D | pest_debugger -g src/elplocalhost: Tue Nov 21 22:59:57 2023

pest_debugger is up to date.
pest_debugger v2.7.5

  --> 38:11
   |
38 | MyFunction()
   |           ^---
   |
   = expected "EOI", "Enum", "Function", "Ident", "Import", "Interface", "StringL
iteral", "TypeBinding", or "ValueBinding"
```

I would expect the FN predicate to be enough to didtinguish the definition from the call? Can someone better than me at this advise please? I'm keen to learn

**Comment by tomtau** [comment]
I reordered `Expression` choices (that's perhaps one main gotcha with PEG that choices are ordered) and it goes through, but it gets stuck at `if`:

```
  --> 45:1
   |
45 | if a {}
   | ^---
   |
   = expected Block, CallArgDeclarations, EOI, Enum, Function, Ident, Import, Interface, or Literal
```

but I don't see `IfElseChain` as a choice put anywhere. Anyway, I hope it helps a bit.

https://pest.rs/?g=N4Ig5gTghgtjURALhABQge0rABAXh2BwGUB5ASRwD8cBRADwAcIBTAZzYEsMA7AKmp0KOAL4AdHhID0UnKgCCAJVoA5ACo41ATVS1iEhs3Zde%2BHAH1gEsQBcAYgFceAYxvceAYSgAbbxJw4AD62ji5uvP5BOAAynDYs0H48AYE45DzxEABmUM4skam0PA4wBWkwjBgQNmUAagicUABG3iwAQpw8ACadYHUNza0AIizO3ghQ4ZI84jwyOGJiCsrqmjp61nOy0eRqtIry0fo8sZk%2BZgACVskBJDYQvacJPmUqJU0JT4llbRgY3l8XjNNr9-oDvGZLJpFABVWhROyHYjw2ZqWHw26YghQsQge4OfIgUQSRFHDGYyFEXE5bxsQnE6bEe6POLPCEU24EQgLECLXG4wQACiZDx4YHB6Uy3AgUXYzigjBYAEoBDR%2BYteUTZiKWWdvJKEtKzNzBQBCQXqvlE1KWzVKwTyFRaJUAagZRi6LCynRYXQ5HOxVJAkmtPIgApteIjPL5GqiuIADNHbcmQAByAWzJoAT3i-vznIsQfoApo8mIHnI5HMAAlaAANcxDcgAcV2wAATCIGU5OM4MJ6CwGizyHKWeVYiWWK1Xaw2m63dm61SBxFqJHKFSwh8OcbzY1OcIKPV6fX7Ujm86le-3PfbZhIemA4gXAzg0wm0wA6L9pgCcaYMpKO5vk%2BL40IKYE2PGIDmLiKoMnY3gYJMr5FpEATASuX7jpKAD8ggsEweFlFhOBEYwEizIw3gOGwQ5vriLqZhIMCdHRaF7gAtCxPAUSBI4AHq4vSEE0RxqRscUbBKgRNDAbMbwwB8EDgpSaQZAiyGobM8yLDsewHEcmzzG0BwqB4NbkCoLYSOQdgCXunBZLxtBkpxQYsLS9KzG5yL2f6jEgF5dLObxACy8hqJZ%2BZBfANjOAAFrx0RYH2Hi8Pc-zqQFhTuXl-kObM5BZLQ3keIlUCdAEb6pc%2BzgZRkmAQjQhisBw7hyRO45tcY7iqjya4MnpYhmY6lnWbZ0zzNZaLWcQ5AeMcADqVl7MQCgePCQU8jBiw1KGKi0MtOxHUB4WoKQigaFiI64pwFRVAdiGKKQ4WBXdIBZJgpTrjws37IiW01Z9nSZDkeSuSoMLvbunnFL9iEqB5PJZCGDIeKQKjEDdt17v2PBsM9swANIsNmADuVRdPRINQnYyOpLldD5SzhVRJF0U1lEAOKED8KFND71MxdV0aKkdivcLOCY9jGgPv9noZI1RNxASGT0QQVw4KaZOU9T9HTpW1aHKgNbyEK5bG%2BYpvm0L%2ByLTBcEgAhxVK9BO6XEQ5Duyrbg2OrNiG0euI4YePssMrvCqwHkdB670iyIss2KPNi3HInOBDLQHjREoUXkFjxz1LR7SdD0YqxdyuJ4D1TDtSYySotmiodN0vRoUGSC4e7DLhSwykJJ7INBmHggR5pNBqC3Zft2KXUlwSbcV2ABG%2BQj-eDzKnc8mP8m9zQqAIJH8gQGAIxjBMUxsGvdkZAkEPbgx3K8-z48H91h6bypA24kN68lGHoWIgqgYbvzjoIXEk5BBFBKN-BIy5Bq8XqA8QYLAL7jGgFMMwRBZY4y6hPaCU8Z7L16LfHgoRXDuE9gQIgDNwGTzkMfHgp9z6jEwZMdwN9BCUKmOkRKhog5dUUCwAOEAeDT0VF1NoyFnAAGthpJzENnXO%2Bc1CF2xiZWQOMoq0HCqoNQxxyCPWqM-b2otrqQJANAiChCrEABpcR4VdFYoaNBJZvUEDqMUalZgoMaC0WeK8cFHn8WgjBV9qErkcSAZxiDF5BI7rpJROi9j6PUBnLYdB6yoGUMQBaRcJAyIwPIoBI9P4wPrn1Xgv9Vy8SPqwFhZ8IlYK4SEi0IAOlCjscQ1u5cO7RKcQhFcSp4IYx8N4VhLTOHR3KbiLpEFBQ9J5LXF2XVeodVMIM2JwyeSjL%2BiIsREiZ4CSDFxAAfD3CBsxeHuH4YI2mwCeTd3Dh-LwvgpnsMidHchtzeDvPZB9b2H9BQAs%2BZfVpszUjFPkS45JCwxANlyXoAp2MQD2JAJ0RgDgbDIExSY6CRB5BsGzC4AAigSAk9i7iijAKIHA30MAwB5F5SivJphgwfrkbc4VswGmyNywgkQvzMAwIqao2YkA4CJrSqimxI6AN5bA5l1wAhflIKgdRWNhXyA8FqlQgoEBgClTK3oozgTTAJkTHAKh5D6JNcyMUEgABuAxAkOtlTwJo-TK4EAAIwuoGD0Zw1LvVzzpf6zYaMcC8r%2BckYAsxo2xqcFQ3ggp7QJokEm7McbBQCN8BgKVVMIDeC6BmhW2a404AuTgGwJzM08ErSmqY6bq3nNrfWxNyRk1hHcHmryyEi1VFLfaGtdbFSEArd2nNzb3BSt5ZI7cDam29rTUqedM7V3xq7TGzdqaeAbtzeWrN07c35sHTgYtI7D2ztMMu09t6eD9oLUOktZab1bufchY90wV37o3YuttHaJ33t3Uej9%2B6gPjqXTunt-6wOPtbWOztJ6EOfvPYWy9w731ocg8hkDsG90tow6%2B69uHiMDowKO9t0HJ2obgxRl9WG33Up4LAFgHqzUQcY9%2BqDKHpgMb7eawTd7ZgiafbiK9XR4ISHE1%2BjAZhJPYdxKx9jimQBdCgM6wkwmiN9qU2%2BsZDa5MYfU1JlTOA2MwG3AQXEmntNGYVs5HAUA6N8W8s5pobmQowc2PFJKOB2oOG8IS6wNgvz6PCm0fYQGKIN3cGFr8ur9WGrPpxsU1GXNnzldMfziUcBND%2BK0KAngBGlOuLYfE24a1xeqdMWwNI6SxaqZs6YIhVxAA#editor

**Reply by davemackintosh** [reply]
That's interesting, I'd read they were ordered and I did try a few combinations but not that configuration. However, yours works- was there any logic to your ordering choice?

Yeah I haven't added it as a choice yet as I was working top to bottom on the kitchen sink.

Thanks for your help!

**Reply by tomtau** [reply]
>  However, yours works- was there any logic to your ordering choice?

I just tried and it worked, but I think the logic is because it came before the choices that expand into having `Ident` in the beginning

**Reply by davemackintosh** [reply]
Ah okay, I was hoping there was something I wasn't understanding. I'll keep poking and prodding. Thanks again for your help!

**Reply by davemackintosh** [reply]
Okay, so I've another interesting question that might be similarly order related but I can't figure out a pattern. So, I've added the branching logic to it but it always fails at `elseif` with expected Block or a paremeter definition. **but** if I remove the `!Keyword ~` from the `IdentConstituents` it works but I get incorrect results. Any thoughts on this? 

**playgrounds**

Broken on `elseif` -> [here](https://pest.rs/?g=N4Ig5gTghgtjURALhABQge0rABAXh2BwGUB5ASRwD8cBRADwAcIBTAZzYEsMA7AKmp0KOAL4AdHhID0UnKgCCAJVoA5ACo41ATVS1iEhs3Zde%2BHAH1gEnDZwBZKABcAxgAtrOAD45yAM1oANmwsAMKuUJySPDbeAGIArjzOjtw8IVABAR5xicmp2TgAMpyOLNBZ0V4%2BPKUQvlDOLAW0PPEwBeQwjBgQjgUAagicUABGASwAQpEAJpFgA0Oj4wAiLM4BCE75POI8MjhiYgrK6po6ehLSsoXkarSK8oX6PMW1GWYAAlaVNsSOEHNXmUMgUVG0RmUgeUChMMBgAlCQTtLjxYfDEQEzJZNIoAKq0KqxR7EAm7NR4gm2KkEbFiED-eJNECiCREp6UqlYoh0%2BpBJksqJ-AE8MAYznUmzckCHOl0wQACiFgJKwIC5BqZW4ECq7GcUEYLAAlAIaLLDtLmbslSKMeralqzIQcPKAITys0y5neD0Ww2CeQqLSGgDUAqM0xYvkiLGm4vFNKlki9BxAEDl3vp6ZTMvNVTpAAYsz6iyAAORy3YjACepTjddsCZT9DlNHkxBC5HI5gAErQABrmZbkADit2AACYRALEpxnBgI-X4xYpfEWymrMzW%2B3Oz3%2B4OR7dQ6aQOJLRJdfqWIul7TpTnN87w5Ho7HvNXa94Z3OI37dhJZmAJT1o2pb5qWAB04GlgAnKWAp2tejYAUBNDyshjh5iA5h0saAqxAEGBOMBy4eDYCHHuBa52gA-IILBMNRHQ1HRTASLsjABPEbCLo2dLBhWEgwJEXHEbeAC0Ak8PRjCIcuOAAHp0vyqEcSJ3hCa0bCGrRNAIbsYIwBCEBio2CFxARRG7Pshw3HcDxPCi%2BwTA8KghN25AqMOEjkLEsm3pwviSbQ7KiVKLB8kF7I%2BXGvEgOFwQBZJdjyGobl1rF8AuO4Z4vFgs4hLw-zwp8RDRd4wUkmVdAhbsfiBMEMVOoUeXOAVNSYJiNCGKwHCpIIEwEc4ADW8H%2BHyYQRJUBBOnVfKhrsDhZSEGxcVejpSpRIA6c65ARsx3gYn6NCoAgLA8PIEBgKs6ybCkvBsNtdJiQAfGui1uIoLCOPEEDRMeAA0SVOB9X0-dEBBEN1xh9d4A0YMNArva46WSvYKVpahUO9aYB0quUR3rm9wOuMtUCrSaKangK1liM5AZuR5XlRPsHnkh5xDkCEzwAOruXcxAKCEBKxSmmGHH0yYqLQ3M3FL8F2KgpCKBoEr%2BV0PQS3hiikHYjW3r4mDtDlrP3ESQs2LFkS1PUjRBSouK6zeYWtEbeEqKFKa%2BEmAohKQKjECrqtSnOPBsJruwANIsFWADuPTTNxFtybE7veFVFW0OnIXeMlqXdlUJuKGbBLlfbutpwrSsaHE2vlzgvv%2Bxof48LtZ2OG1YclIyNTcQQXw4C6Uex-H3Fbh2XaPKg3byAqbbj%2BYk-T2X9yc5h2EgLhtV7Rh14lT428dyk31t6Pzp0ptgitzUh9dyfm9XAcYis4o7Oc88D-LLQISFEoKXkH7zxBicUmDMOYKMpR4DXFjEw0QyRVgNFMHgswRTESlEgKi29EYsEMmUXeScNoYLboINQ8CQFILmNtIBjJEHILALRXYLQ2h2GwUZHiTpz6EOYsdU651LrXQ2NAO6od6HeQ1HUBoa1UE%2BHUKbeQ5tdKYOPBuQQzCcEQApnSKmDCXZ4IbE6VQDtL6KMJg%2BRhMBVFGSPJTSSgwARLBYPw26fUIb1wAWobaV8MI0BIQg0BIoRE8ASEkIRu8XEpyMUQ7hrBeFXTWAIrY91tpBLyLwdUrhNSOAeoIT631fo%2BJYNtOGCNdg5LBvkthUoXqcIwlZWQhxP7f1-mof%2B-tHKyADilWgdhVBqGeJ0bovQKk%2BErsrQQdJlGoU8WMkAgMtohmmVTGgsRa6CGtKKPG7xdi2OGGMMhtCzBEHlNs%2BxjjBF9QBnSai8yaBUL2WA2pj8Ol3G6eod%2BexZD9lQMoYgHMAESCKSNXRByTGCGgakDRJ5JInWiRdWJN0zn3WBe6EAyKFRTO8aQmhYCLlzIhYaHCPsMgBFhachJod8EplRaheU6KUyQI3ttMFpgcVXLxZJZJQi0kZMTnotB1TBDpEyCSuJTjEl4VyEIwVmJwH70ic6KVwr4Vku4rDQaQ15kPMOJ875vzWk8BmSASIjB4iOGQIa9WgyiDyDYFWJIABFRkjJ-okH%2BPcnABsMAwBTOFRgsoohWzKDbK8dgqx2kDRIwgHhwLMAwAaXoVYkA4DDsKeYyIohnTaPYKsZjI2VHAqQVAzS-ZRrkUWlQ8oEBgETcmuY%2BK00SBDmHHAKh5DdOra6kUEgABuixdntpTRIEYfiwBmAAIzdsWLMZwzqh3kJQQQcdPAW1tqTR2kdC7xwAGYUReyzRyvqwBdi7pDfu3g8o-SHokMeqsp6eDynSZkDAia44QACNMC9zdr23pwC9HAjhSGECPdEE9ErUjnp-c9P9AHL08C-aBs9D6CLPp6G%2Bv0v7-0GkA1e4DN74PRG%2BDYRtGEeCwCvAutiO6cO3sTSG8pMG4PBLA4aGjuHGOmHo1RvDLHb3gY43uvD97wpIZwC%2B1D3G8NYdg5xtjd7ENPpEyh994mZOCcfR%2Byj-GZMsfKeh6DQHNMpLvcxgzITdOYb4yBrTJmmMQag%2BZ-TlnDOqeE6JpT1nTBmavBZ1jTm5PIdfW5xzQjnMYDQ5BjDXmHM%2BeC35hTAXnUkZgCwfttblO%2BaE6F2zEXJMSCC0x3L0WD0LUK2eukrmcIFZ43JswZXFN0gS6RmrIBphQC7UyOteXSsgHKxvSTnXZMZaaz1hrSWmstbawSmDEgAo4CgJJ%2BKLAZsjCdLsBbOWeCZTcDgHq8QAgYW%2BGIRw4Ful2AmPcWz0keowMBpII7paWkVsuilkUYXZuXRu83TbyMRhwnGFANI6SEYHccAyK8v7LvQ14Ddw7vIGrg6YFd1IH2JAniAA#editor)
Works but I get if, else and elseif as `VariableDeclaration`s instead of `IfElse` -> [here](https://pest.rs/?g=N4Ig5gTghgtjURALhABQge0rABAXh2BwGUB5ASRwD8cBRADwAcIBTAZzYEsMA7AKmp0KOAL4AdHhID0UnKgCCAJVoA5ACo41ATVS1iEhs3Zde%2BHAH1gEnDZwBZKABcAxgAtrOAD45yAM1oANmwsAMKuUJySPDbeAGIArjzOjtw8IVABAR5xicmp2TgAMpyOLNBZ0V4%2BPKUQvlDOLAW0PPEwBeQwjBgQjgUAagicUABGASwAQpEAJpFgA0Oj4wAiLM4BCE75POI8MjhiYgrK6po6ehLSsoXkarSK8oX6PMW1GWYAAlaVNsSOEHNXmUMgUVG0RmUgeUChMMBgAlCQTtLjxYfDEQEzJZNIoAKq0KqxR7EAm7NR4gm2KkEbFiED-eJNECiCREp6UqlYoh0%2BpBJksqJ-AE8MAYznUmzckCHOl0wQACiFgJKwIC5BqZW4ECq7GcUEYLAAlAIaLLDtLmbslSKMeralqzIQcPKAITys0y5neD0Ww2CeQqLSGgDUAqM0xYvkiLGm4vFNKlki9BxAEDl3vp6ZTMvNVTpAAYsz6iyAAORy3YjACepTjddsCZT9DlNHkxBC5HI5gAErQABrmZbkADit2AACYRALEpxnBgI-X4xYpfEWymrMzW%2B3Oz3%2B4OR7dQ6aQOJLRJdfqWIul7TpTnN87w5Ho7HvNXa94Z3OI37dhJZmAJT1o2pb5qWAB04GlgAnKWAp2tejYAUBNDyshjh5iA5h0saAqxAEGBOMBy4eDYCHHuBa52gA-IILBMNRHQ1HRTASLsjABPEbCLo2dLBhWEgwJEXHEbeAC0Ak8PRjCIcuOAAHp0vyqEcSJ3hCa0bCGrRNAIbsYIwBCEBio2CFxARRG7Pshw3HcDxPCi%2BwTA8KghN25AqMOEjkLEsm3pwviSbQ7KiVKLB8kF7I%2BXGvEgOFwQBZJdjyGobl1rF8AuO4Z4vFgs4hLw-zwp8RDRd4wUkmVdAhbsfiBMEMVOoUeXOAVNSYJiNCGKwHCpIIEwEc4ADW8H%2BHyYQRJUBBOnVfKhrsDhZSEGxcVejpSpRIA6c65ARsx3gYn6NCoAgLA8PIEBgKs6ybCkvBsNtdJiQAfGui1uIoLCOPEEDRMeAA0SVOB9X0-dEBBEN1xh9d4A0YMNArva46WSvYKVpahUO9aYB0quUR3rm9wOuMtUCrSaKangK1liM5AZuR5XlRPsHnkh5xDkCEzwAOruXcxAKCEBKxSmmGHH0yYqLQ3M3FL8F2KgpCKBoEr%2BV0PQS3hiikHYjW3r4mDtDlrP3ESQs2LFkS1PUjRBSouK6zeYWtEbeEqKFKa%2BEmAohKQKjECrqtSnOPBsJruwANIsFWADuPTTNxFtybE7veFVFW0OnIXeMlqXdlUJuKGbBLlfbutpwrSsaHE2vlzgvv%2Bxof48LtZ2OG1YclIyNTcQQXw4G2HZdo8qDdvICqDzuI9j2X9yc5h2EgLhtV7Rh14lT4q8dyk31t9xqF0ptgitzU29d3vy9XAcYis4o7Oc88V-LLQISFEoKXkH7zyDJxkwzHMKMpR4DXFjEw0QyRVgNFMHgswRTESlEgKiq9EYsEMmUdeScNpILboINQkC-4wLmNtH%2BjJoGwLALRXYLQ2h2FQUZHiTpD7YOYsdU651LrXQ2NAO6odKHeQ1HUBoa14E%2BHUKbeQ5tdLIOPBuQQtC0EQApnSKmVCXYYIbE6VQDtj7SMJg%2BahMB5FGSPJTSSgwARLBYJw26fUIb1y-mobaJ8MI0DwVA-%2BIo%2BE8ASEkHh687Epx0Tg1hrB2FXTWFwrY91to%2BLyLwdUrhNSOAeoIT631fpuJYNtOGCNdhpLBpkhhUoXrMIwlZWQhxn6v3fmoT%2B-tHKyADilWgdhVBqGeJ0bovQik%2BErsrQQdJZGoWcQMkAgMtohlGVTGgsRa6CGtKKPG7xdjmOGGMAh5CzBEHlKsyx1juF9QBnSaikyaAkI2QA8p18ml3FaeoR%2BexZD9lQMoYgHMv4SBySNdRWy9GCFAakJRJ5JInVCRdcJN0Dn3V%2Be6EAsKFQjNcfgshACjkTKBYaHCPsMgBHBfsqJodMEpnhaheUiKUzAKXttAFpg0UnIxZJWJPCElJMThohBpTBDpEyHiiJNjol4VyDw7lmJAGb2Cc6EVvLIUEu4rDQaQ1JlXMOM8157z6k8DGSASIjB4iOGQNq9W3SiDyDYFWJIABFRkjJ-okH%2BJcnABsMAwBTOFRgsoohWzKDbK8dgqx2m9UIwgHhwLMAwAaXoVYkA4DDsKeYyIohnTaPYKsBjg2VHAqQVAtS-YhokTmlQ8oEBgGjbGuYmKE0SBDmHHAKh5CtNLfakUEgABuix1mNrjRIEYHiwBmAAIytsWLMZwtqe2ELgQQQdPA60NpjU2vtU7xwAGYURexTUyvqwBdjrr9Zu3g8o-TbokLuqs%2B6eDykSZkDA0a44QACNMI9zdT3npwC9HAjh8GEB3dEPdQrUiHrfc9D9X7j08Bff%2Bg9V6CK3p6A%2Bv077P0Gm-Se39Z7IPRG%2BDYatGEeCwCvFOtia60PnujX6wpYGIO%2BIA4aMj6HqOmEoyRjDdHz2AaYxujDl7wowZwHe%2BDrGMMofA8xhjF7oM3r43Bx9gmxPcevU%2B4jnGxN0cKYh0DP7lNxIvbRrTfj1PIY43%2BlTemaNAZA4ZzTxntPyd4-xmTpnTAGavEZ%2BjNmJOwfvQ56zPDbMYAQ8BpDLmrNud8x5qTXnbV4ZgCwTt5bZPuZ4-58zQXhMSB8zR9LoWt0LWyweuk9mcJZbYxJswBXpN0ii-hsrIBphQBbUyCtGX8sgEK0vYTzXxNJZq21qrMWat1Ya1isDEgAo4CgMJ%2BKLAxsjCdLsKbaWeCZTcDgHq8QAgYW%2BGIRw4FWl2AmPccz0kepgMBpIHb%2Ba6lFsunFkUAXxuXTO83ZbyMRhwnGFANIiSEZbccAyK877jvQ14Gd7bvIGqA6YCd1IT2JAniAA#editor)

**Reply by tomtau** [reply]
> Okay, so I've another interesting question that might be similarly order related but I can't figure out a pattern. So, I've added the branching logic to it but it always fails at elseif with expected Block or a paremeter definition. but if I remove the !Keyword ~ from the IdentConstituents it works but I get incorrect results. Any thoughts on this?

I think it may be because of the unambiguous greedy nature, `!Keyword` may consume/match some characters that you expect to be matched by `ASCII_ALPHA`, plus there are many `Expression*` nestings... you can perhaps try to load up all rules in a debugger to step through how it got there

**Comment by davemackintosh** [comment]
I'm going to close this as I've written my own lexer and parser for my language. Thanks for your help though.
